File: frontend/src/orbitalMotion.ts
import type { Scenario, TargetSpec, TrajectoryPoint } from "./types";
import { elementsFromState, propagateElements, sampleOrbitPath } from "./kepler";

const MU = 398600;
const EARTH_SPIN_RAD_S = (2 * Math.PI) / 86164;

export interface TargetMotion {
  orbit_radius_km: number;
  initial_angle_rad: number;
  angular_velocity_rad_s: number;
  spin_rad_s: number;
  kind: "satellite" | "moon";
  revolution_period_s?: number | null;
  orbit_type: "circular" | "elliptical";
  semi_major_axis_km: number;
  eccentricity: number;
}

function targetElements(target: TargetSpec): ReturnType<typeof elementsFromState> | null {
  if (target.velocity) {
    return elementsFromState(target.position, target.velocity);
  }
  if (target.orbit_elements) {
    const o = target.orbit_elements;
    return {
      semi_major_axis_km: o.semi_major_axis_km,
      eccentricity: o.eccentricity,
      argument_of_periapsis_rad: o.argument_of_periapsis_rad,
      true_anomaly_at_t0_rad: o.true_anomaly_at_t0_rad,
      mean_motion_rad_s: o.mean_motion_rad_s,
    };
  }
  if (target.semi_major_axis_km != null) {
    const a = target.semi_major_axis_km;
    const e = target.eccentricity ?? 0;
    const argp = target.argument_of_periapsis_rad ?? 0;
    const [x, y] = target.position;
    const nu0 =
      target.true_anomaly_at_t0_rad ?? Math.atan2(y, x) - argp;
    return {
      semi_major_axis_km: a,
      eccentricity: e,
      argument_of_periapsis_rad: argp,
      true_anomaly_at_t0_rad: nu0,
      mean_motion_rad_s: Math.sqrt(MU / (a * a * a)),
    };
  }
  return null;
}

export function getTargetMotion(target: TargetSpec): TargetMotion {
  const x = target.position[0];
  const y = target.position[1];
  const el = targetElements(target);
  const r = target.orbit_radius_km ?? Math.hypot(x, y);
  const omega =
    el?.mean_motion_rad_s ??
    target.angular_velocity_rad_s ??
    Math.sqrt(MU / (r * r * r));
  return {
    orbit_radius_km: r,
    initial_angle_rad: target.initial_angle_rad ?? Math.atan2(y, x),
    angular_velocity_rad_s: omega,
    spin_rad_s: target.spin_rad_s ?? 0.12,
    kind: target.kind ?? "satellite",
    revolution_period_s: target.revolution_period_s ?? (2 * Math.PI) / omega,
    orbit_type: (target.orbit_type as "circular" | "elliptical") ?? (el && el.eccentricity > 1e-4 ? "elliptical" : "circular"),
    semi_major_axis_km: el?.semi_major_axis_km ?? r,
    eccentricity: el?.eccentricity ?? 0,
  };
}

export function targetOrbitPath(target: TargetSpec): [number, number][] {
  const el = targetElements(target);
  if (el) return sampleOrbitPath(el);
  const m = getTargetMotion(target);
  const pts: [number, number][] = [];
  for (let i = 0; i <= 90; i++) {
    const theta = m.initial_angle_rad + (2 * Math.PI * i) / 90;
    pts.push([m.orbit_radius_km * Math.cos(theta), m.orbit_radius_km * Math.sin(theta)]);
  }
  return pts;
}

export function targetPositionAt(target: TargetSpec, t_s: number): [number, number] {
  const el = targetElements(target);
  if (el) return propagateElements(el, t_s);
  const m = getTargetMotion(target);
  const theta = m.initial_angle_rad + m.angular_velocity_rad_s * t_s;
  return [m.orbit_radius_km * Math.cos(theta), m.orbit_radius_km * Math.sin(theta)];
}

export function earthSpinAngle(t_s: number, earthSpinRadS?: number): number {
  return (earthSpinRadS ?? EARTH_SPIN_RAD_S) * t_s;
}

export function bodySpinAngle(target: TargetSpec, t_s: number): number {
  return getTargetMotion(target).spin_rad_s * t_s;
}

export function targetPointAtTime(
  scoreTargetTraj: TrajectoryPoint[] | undefined,
  target: TargetSpec,
  t_s: number,
): [number, number] {
  if (scoreTargetTraj?.length) {
    const nearest = scoreTargetTraj.reduce((best, p) =>
      Math.abs(p.t_s - t_s) < Math.abs(best.t_s - t_s) ? p : best,
    );
    return nearest.position;
  }
  return targetPositionAt(target, t_s);
}

export function chaserCoastPath(scenario: Scenario): [number, number][] {
  return sampleOrbitPath(
    elementsFromState(scenario.spacecraft.position, scenario.spacecraft.velocity),
  );
}

File: orbital_planner/orbital_motion.py
"""Moving targets: circular + elliptical Kepler orbits, Earth spin."""

from __future__ import annotations

import math

from orbital_planner.constants import (
    DEFAULT_BODY_SPIN_RAD_S,
    EARTH_SPIN_RAD_S,
    MOON_RADIUS_KM,
    MU_EARTH_KM3_S2,
    MU_MOON_KM3_S2,
)
from orbital_planner.kepler2d import (
    OrbitElements,
    elements_from_state,
    mean_motion,
    propagate_elements,
    sample_orbit_positions,
    state_from_elements,
)
from orbital_planner.schemas import TargetSpec, TrajectoryPoint


def circular_mean_motion(radius_km: float, mu: float = MU_EARTH_KM3_S2) -> float:
    r = max(radius_km, 1.0)
    return math.sqrt(mu / r**3)


def _elements_out_to_dataclass(out: "OrbitElementsOut") -> OrbitElements:
    from orbital_planner.schemas import OrbitElementsOut

    assert isinstance(out, OrbitElementsOut)
    return OrbitElements(
        semi_major_axis_km=out.semi_major_axis_km,
        eccentricity=out.eccentricity,
        argument_of_periapsis_rad=out.argument_of_periapsis_rad,
        true_anomaly_at_t0_rad=out.true_anomaly_at_t0_rad,
        mean_motion_rad_s=out.mean_motion_rad_s,
        period_s=out.period_s,
        periapsis_km=out.periapsis_km,
        apoapsis_km=out.apoapsis_km,
        specific_angular_momentum=out.specific_angular_momentum,
        specific_energy_km2_s2=out.specific_energy_km2_s2,
    )


def target_elements(target: TargetSpec) -> OrbitElements:
    tg = enrich_target(target)
    if tg.orbit_elements is None:
        raise RuntimeError("Target orbit elements not resolved")
    return _elements_out_to_dataclass(tg.orbit_elements)


def target_body_mu(target: TargetSpec) -> float:
    """Gravitational parameter for third-body pull (0 for massless satellites)."""
    if target.gravitational_parameter_km3_s2 is not None:
        return target.gravitational_parameter_km3_s2
    if target.kind == "moon":
        return MU_MOON_KM3_S2
    return 0.0


def target_body_radius_km(target: TargetSpec) -> float:
    if target.body_radius_km is not None:
        return target.body_radius_km
    if target.kind == "moon":
        return MOON_RADIUS_KM
    return 0.0


def enrich_target(target: TargetSpec) -> TargetSpec:
    """Fill orbit fields; prefer elliptical elements from r,v or a,e,ν."""
    x, y = target.position
    r0 = math.hypot(x, y)

    elements: OrbitElements | None = None

    if target.velocity is not None:
        elements = elements_from_state(target.position, target.velocity)

    if elements is None and target.semi_major_axis_km is not None:
        a = target.semi_major_axis_km
        e = max(0.0, min(target.eccentricity, 0.95))
        argp = target.argument_of_periapsis_rad
        nu0 = target.true_anomaly_at_t0_rad if target.true_anomaly_at_t0_rad is not None else math.atan2(y, x) - argp
        n = mean_motion(a)
        elements = OrbitElements(
            semi_major_axis_km=a,
            eccentricity=e,
            argument_of_periapsis_rad=argp,
            true_anomaly_at_t0_rad=nu0,
            mean_motion_rad_s=n,
            period_s=2.0 * math.pi / n,
            periapsis_km=a * (1.0 - e),
            apoapsis_km=a * (1.0 + e),
            specific_angular_momentum=math.sqrt(MU_EARTH_KM3_S2 * a * (1.0 - e * e)),
            specific_energy_km2_s2=-MU_EARTH_KM3_S2 / (2.0 * a),
        )

    if elements is None:
        # Circular fallback
        omega = target.angular_velocity_rad_s or circular_mean_motion(r0)
        elements = OrbitElements(
            semi_major_axis_km=r0,
            eccentricity=0.0,
            argument_of_periapsis_rad=target.initial_angle_rad or math.atan2(y, x),
            true_anomaly_at_t0_rad=0.0,
            mean_motion_rad_s=omega,
            period_s=2.0 * math.pi / omega,
            periapsis_km=r0,
            apoapsis_km=r0,
            specific_angular_momentum=math.sqrt(MU_EARTH_KM3_S2 * r0),
            specific_energy_km2_s2=-MU_EARTH_KM3_S2 / (2.0 * r0),
        )

    from orbital_planner.schemas import OrbitElementsOut

    orbit_type = "circular" if elements.eccentricity < 1e-4 else "elliptical"
    out = OrbitElementsOut(
        semi_major_axis_km=elements.semi_major_axis_km,
        eccentricity=elements.eccentricity,
        argument_of_periapsis_rad=elements.argument_of_periapsis_rad,
        true_anomaly_at_t0_rad=elements.true_anomaly_at_t0_rad,
        mean_motion_rad_s=elements.mean_motion_rad_s,
        period_s=elements.period_s,
        periapsis_km=elements.periapsis_km,
        apoapsis_km=elements.apoapsis_km,
        specific_angular_momentum=elements.specific_angular_momentum,
        specific_energy_km2_s2=elements.specific_energy_km2_s2,
    )
    return target.model_copy(
        update={
            "orbit_type": orbit_type,
            "orbit_elements": out,
            "semi_major_axis_km": elements.semi_major_axis_km,
            "eccentricity": elements.eccentricity,
            "argument_of_periapsis_rad": elements.argument_of_periapsis_rad,
            "true_anomaly_at_t0_rad": elements.true_anomaly_at_t0_rad,
            "angular_velocity_rad_s": elements.mean_motion_rad_s,
            "revolution_period_s": elements.period_s,
            "orbit_radius_km": r0,
            "initial_angle_rad": math.atan2(y, x),
            "gravitational_parameter_km3_s2": target_body_mu(target),
            "body_radius_km": target_body_radius_km(target),
        },
    )


def target_position_at(target: TargetSpec, t_s: float) -> tuple[float, float]:
    el = target_elements(target)
    pos, _, _ = propagate_elements(el, t_s)
    return pos


def target_velocity_at(target: TargetSpec, t_s: float) -> tuple[float, float]:
    el = target_elements(target)
    _, vel, _ = propagate_elements(el, t_s)
    return vel


def target_true_anomaly_at(target: TargetSpec, t_s: float) -> float:
    el = target_elements(target)
    _, _, nu = propagate_elements(el, t_s)
    return nu


def sample_target_trajectory(
    target: TargetSpec,
    time_limit_s: float,
    *,
    sample_every_s: float = 30.0,
) -> list[TrajectoryPoint]:
    points: list[TrajectoryPoint] = []
    t = 0.0
    while t <= time_limit_s + 1e-9:
        pos = target_position_at(target, t)
        vel = target_velocity_at(target, t)
        points.append(TrajectoryPoint(t_s=t, position=pos, velocity=vel))
        t += sample_every_s
    return points


def target_orbit_path(target: TargetSpec, n_points: int = 120) -> list[tuple[float, float]]:
    return sample_orbit_positions(target_elements(target), n_points=n_points)


def earth_spin_angle(t_s: float) -> float:
    return EARTH_SPIN_RAD_S * t_s


def body_spin_angle(target: TargetSpec, t_s: float) -> float:
    return target.spin_rad_s * t_s


def spacecraft_elements(scenario_spacecraft_position, scenario_spacecraft_velocity) -> OrbitElements:
    return elements_from_state(scenario_spacecraft_position, scenario_spacecraft_velocity)
